### PR TITLE
AWS kube-up: Reboot support with ephemeral storage

### LINKED ISSUE #21157: AWS: Don't try to reinitialize ephemeral storage on reboot
On AWS we (will) now run our bootstrap script on every boot.  The function to initialize ephemeral volumes doesn't tolerate being re-run, and we probably shouldn't re-run it anyway for speed.

This does preserve the contents, which may be a good thing to save user's data (it doesn't belong on ephemeral drives, but...).  On the other hand, if the ephemeral drives are corrupted, we lose the opportunity to reformat them and recover.  Arguably, if the ephemeral storage is corrupted the right thing to do is to terminate the instance entirely.

### PR DESCRIPTION
**Pull request by justinsb**
If the ephemeral volume is present and mounted, don't try to reinitialize
them.

Don't block the boot if the ephemeral volume is corrupt / missing -
this enables us to cope with a stop/start & presumably also corruption.
In this case, we'll reformat the ephemeral storage.

Fix #21157

### CODE REVIEW

File: cluster/aws/templates/format-disks.sh [hunk 1]
```diff
@@ -183,7 +193,9 @@ if [[ -n "${move_docker}" ]]; then
     mv /var/lib/docker ${move_docker}/
   fi
   mkdir -p ${move_docker}/docker
-  ln -s ${move_docker}/docker /var/lib/docker
+  if [[ ! -e /var/lib/docker ]]; then
```
Comment (brendandburns) [inline cluster/aws/templates/format-disks.sh]: Is there a race here?  How can that directory ever exist, we just moved it away above on 193

File: cluster/aws/templates/format-disks.sh [hunk 2]
```diff
@@ -195,7 +207,9 @@ if [[ -n "${move_kubelet}" ]]; then
     mv /var/lib/kubelet ${move_kubelet}/
   fi
   mkdir -p ${move_kubelet}/kubelet
-  ln -s ${move_kubelet}/kubelet /var/lib/kubelet
+  if [[ ! -e /var/lib/kubelet ]]; then
```
Comment (brendandburns) [inline cluster/aws/templates/format-disks.sh]: as above wrt to how this directory can exist given that we must moved it.

File: cluster/aws/templates/format-disks.sh (hunk 1 above)
Comment (justinsb) [inline cluster/aws/templates/format-disks.sh]: Yeah, messy... I'll add a comment to explain.

We want `/var/lib/docker` to be a symlink to `/mnt/ephemeral/docker`.  Line 193 only fires if it is a directory (`-d`).  If you stop and start your instance, that corrupts /mnt/ephemeral (ephemeral disks) but the root volume is usually persisted (if on EBS).

So the problem we're solving here is that /var/lib/docker may already be a symlink to /mnt/ephemeral/docker, but /mnt/ephemeral may have been wiped.

**Comment by justinsb** [comment]
Added comments to explain the logic here.

I would love to replace this script entirely, either with cloud-init if it serves our purpose or a little go program if not.  There's a number of issues here: the logic is complicated; I don't think we recover from all cases of corrupted local storage; we don't mount all the ephemeral disks; we should mount a persistent EBS volume if the user specifies that instead; we don't use e.g. btrfs if there are no ephemeral volumes.  I think we have issues for most of those issues, but I'll open an issue to address this in 1.3.

**Comment by brendandburns** [comment]
Thanks for the explanation.

**Pull request merged**